keps/sig-storage/20191125-bucket-provisioning.md

```diff
@@ -0,0 +1,420 @@
+1. `bucketAccessName`: name of the bound cluster-scoped `BucketAccess` instance
+
+#### BucketAccess
+
+A cluster-scoped administrative API which encapsulates fields from the BucketAccessRequest and the BucketAccessClass.  The purpose of the API is to server as communication path between provisioner and a central COSI controller.  In Greenfield, the COSI controller will create BucketAccess instances for new BucketAccessRequests.  The relationship between a BucketAccessRequests and a BucketAccess should be 1:1.  
+
+```yaml
+apiVersion: cosi.io/v1alpha1
+kind: BucketAccess
+metadata: 
+  name: [1]
+  labels:
+    cosi.io/provisioner: [2]
+  finalizers:
+  - cosi.io/finalizer [3]
+ spec:
+  bucketAccessRequestName: [4]
+  bucketAccessRequestNamespace: [5]
+  serviceAccountName: [6]
+  keySecretName: [7]
+  provisioner: [8]
+  parameters: [9]
+```
+
+1. `name`: For Greenfield, generated in the pattern of `bucketAccessRequest-<my-bucketAccessRequest>-<my-bucketAccessRequest-namespace>`. 
+1. `labels`: should be added by controller.  Key’s value should be the provisioner name. Characters that do not adhere to [Kubernetes label conventions](https://kubernetes.io/docs/concepts/overview/working-with-objects/labels/#syntax-and-character-set) will be converted to ‘-’.
+1. `finalizers`: should be added by controller to defer `BucketAccess` deletion until backend deletion ops succeed.
+1. `bucketAccessRequestName`/`bucketAccessRequestNamespace`: name and namespace of the bound `BucketAccessRequest`
+1. `serviceAccountName`: name of the Kubernetes ServiceAccount specified by the `BucketAccessRequest`.  Undefined when the `BucketAccessRequest.accessSecretName` is defined.
+1. `keySecretName`: name of the *provisioner* generated `Secret` containing access credentials. This `Secret` exists in the provisioner’s namespace and must be copied to the app namespace by the COSI controller.
+1. `provisioner`:  name of the provisioner that should handle this `BucketAccess` instance.  Copied from the `BucketAccessClass`.
+1. `parameters`:  A map of string, string key values.  Allows admins to control user and access provisioning by setting provisioner key-values.
+
+#### BucketAccessClass
+
+A cluster-scoped API providing a way for admins to specify policies that may be used to access buckets.  These are always applicable in Greenfield, where access is dynamically granted, and only sometimes applicable in Brownfield, where a user identity may already exist, but not yet have access to a bucket.  In this case, a BucketAccessRequest will still specify the BucketAccessClass in order to determine which actions it is granted to a defined bucket.
+
+```yaml
+apiVersion: cosi.io/v1alpha1
+kind: BucketAccessClass
+metadata: 
+  name:
+provisioner: [1]
+supportedProtocols: [2]
+policyActions: [3]
+  allow:
+  - "*"
+  deny:
+  - "*"
+parameters: [4]
+```
```

Comment:
@jeffvance @copejon what about scope of quota(size, no of objects) here?

Reply:
As with the[ bucket provisioning lib](https://github.com/kube-object-storage/lib-bucket-provisioner), the `parameters` section can be used to pass on quota information. It's unclear to me how quota will work within  object stores existing outside of k8s but referenced by resources in a k8s cluster. 
For now, we can limit the number of instances of these bucket-related CRs, eg. limit the number of `BucketRequest`s.